### PR TITLE
MOTECH-2420 Add task action: Enroll Patient in Program

### PR DESCRIPTION
**Pull request by martokarski**
I added new task action Create Program Enrollment. Program Enrollment informs about patient being enrolled in program and keeps history of patient states.

This action is only available with 1.10 and higher version of openmrs. I tested it on OpenMRS 1.11.6 with rest 2.14 module.

I also moved .json files to proper directories to clean up.

### CODE REVIEW

File: openmrs-19/src/main/java/org/motechproject/openmrs19/resource/ProgramEnrollmentResource.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package org.motechproject.openmrs19.resource;
+
+import org.motechproject.openmrs19.config.Config;
+import org.motechproject.openmrs19.domain.ProgramEnrollment;
+
+/**
+ * Initerface for programs enrollment management.
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/resource/ProgramEnrollmentResource.java]: misspelling in Interface

File: openmrs-19/src/main/java/org/motechproject/openmrs19/resource/ProgramEnrollmentResource.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+ */
+public interface ProgramEnrollmentResource {
+
+    ProgramEnrollment createProgramEnrollment(Config config, ProgramEnrollment programEnrollment);
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/resource/ProgramEnrollmentResource.java:19]: javadoc here

File: openmrs-19/src/main/java/org/motechproject/openmrs19/resource/impl/ProgramEnrollmentResourceImpl.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    public ProgramEnrollment createProgramEnrollment(Config config, ProgramEnrollment programEnrollment) {
+        String requestJson = buildGsonWithAdapters().toJson(programEnrollment);
+        String responseJson = postForJson(config, requestJson, "/programenrollment");
+        return (ProgramEnrollment) JsonUtils.readJson(responseJson, ProgramEnrollment.class);
+    }
+
+    private Gson buildGsonWithAdapters() {
+        return new GsonBuilder()
+                .setDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ")
+                .registerTypeAdapter(Patient.class, new Patient.PatientSerializer())
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/resource/impl/ProgramEnrollmentResourceImpl.java:35]: what about location ? we do not need LocationSerializer ?

Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/resource/impl/ProgramEnrollmentResourceImpl.java:35]: I see you added expose annotation to ProgramEnrollment model however here you do not use 'excludefieldswithoutexposeannotation' option.

File: openmrs-19/src/main/java/org/motechproject/openmrs19/tasks/impl/OpenMRSActionProxyServiceImpl.java [hunk 4]
```diff
@@ -102,6 +106,24 @@ public void updatePerson(String personUuid, String givenName, String middleName,
...
+        Patient patient = new Patient();
+        patient.setUuid(patientUuid);
+
+        Program program = new Program();
+        program.setUuid(programUuid);
+
+        ProgramEnrollment programEnrollment = new ProgramEnrollment();
+        programEnrollment.setPatient(patient);
+        programEnrollment.setProgram(program);
+        programEnrollment.setDateEnrolled(Objects.nonNull(dateEnrolled) ? dateEnrolled.toDate() : null);
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/tasks/impl/OpenMRSActionProxyServiceImpl.java]: dateEnrolled is required, so this field never be null, am I right ?

File: openmrs-19/src/test/resources/json/programEnrollment/program-enrollment-create.json [hunk 5]
```diff
@@ -0,0 +1,6 @@
+{
+  "program": "37677597-27e3-4613-9514-e4d2b7b89cfd",
```
Comment (kmadej) [inline openmrs-19/src/test/resources/json/programEnrollment/program-enrollment-create.json]: why don't you test adding location or states to program enrollment ?

File: openmrs-19/src/main/java/org/motechproject/openmrs19/tasks/OpenMRSActionProxyService.java [hunk 6]
```diff
@@ -100,4 +100,14 @@ void updatePerson(String personUuid, String givenName, String middleName, String
...
+    /**
+     * Creates a program enrollment with the given params.
+     * The required fields are: {@code patientUuid}, {@code programUuid}, {@code dateEnrolled}.
+     *
+     * @param patientUuid the patient uuid
+     * @param programUuid the program uuid
+     * @param dateEnrolled the program enrollment date
+     * @param dateCompleted the program completed date
+     */
+    void createProgramEnrollment(String patientUuid, String programUuid, DateTime dateEnrolled, DateTime dateCompleted);
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/tasks/OpenMRSActionProxyService.java]: I see that ProgramEnrollment has also states and location, what about it ? Can you write why don't you use this fields in task action.

Comment (martokarski) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/tasks/OpenMRSActionProxyService.java]: I added location parameter. States of program enrollment will be updated by another action.

File: openmrs-19/src/main/java/org/motechproject/openmrs19/tasks/OpenMRSActionProxyService.java [hunk 7]
```diff
@@ -118,5 +118,6 @@ void updatePerson(String configName, String personUuid, String givenName, String
      * @param dateEnrolled the program enrollment date
      * @param dateCompleted the program completed date
      */
-    void createProgramEnrollment(String patientUuid, String programUuid, DateTime dateEnrolled, DateTime dateCompleted);
+    void createProgramEnrollment(String configName, String patientUuid, String programUuid,
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/tasks/OpenMRSActionProxyService.java:133]: adjust javadoc

File: openmrs-19/src/main/resources/task-channel.json [hunk 8]
```diff
@@ -321,17 +321,17 @@
       "serviceMethod": "createProgramEnrollment",
```
Comment (kmadej) [inline openmrs-19/src/main/resources/task-channel.json]: why do you change this file ?

File: openmrs-19/src/main/java/org/motechproject/openmrs19/tasks/OpenMRSActionProxyService.java [hunk 9]
```diff
@@ -108,4 +108,17 @@ void updatePerson(String configName, String personUuid, String givenName, String
...
+     * Creates a program enrollment with the given params.
+     * The required fields are: {@code patientUuid}, {@code programUuid}, {@code dateEnrolled}.
+     *
+     * @param patientUuid the patient uuid
+     * @param programUuid the program uuid
+     * @param dateEnrolled the program enrollment date
+     * @param dateCompleted the program completed date
+     * @param locationName the name of location
+     */
+    void createProgramEnrollment(String configName, String patientUuid, String programUuid,
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/tasks/OpenMRSActionProxyService.java:133]: adjust javadoc please

**Comment by kmadej** [comment]
@martokarski please resolve conflicts

**Pull request merged**